From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Sun, 18 Jan 2026 03:17:42 -0500
Subject: [PATCH] Lithium: faster hash palette

TODO: Needs to check compatibility with FAWE's //copy **//paste**
data stored in LithiumHashPalette looks like exactly same with in HashMapPalette
And it can work well on WorldEdit, but not on FAWE

This patch is based on the following mixins:
* "net/caffeinemc/mods/lithium/mixin/chunk/palette/StrategyMixin.java"
By: 2No2Name <2No2Name@web.de>
As part of: Lithium (https://github.com/CaffeineMC/lithium)
Licensed under: LGPL-3.0 (https://www.gnu.org/licenses/lgpl-3.0.html)

diff --git a/net/minecraft/world/level/chunk/Strategy.java b/net/minecraft/world/level/chunk/Strategy.java
index 787f5173425f617a89033e5b76256688be36b31d..a08756342f3cdb3a1188b2e6e56c59c4742db0d1 100644
--- a/net/minecraft/world/level/chunk/Strategy.java
+++ b/net/minecraft/world/level/chunk/Strategy.java
@@ -4,6 +4,15 @@ import net.minecraft.core.IdMap;
 import net.minecraft.util.Mth;
 
 public abstract class Strategy<T> {
+    // Leaf start - Lithium - faster hash palette
+    private static final Palette.Factory HASH = net.caffeinemc.mods.lithium.common.world.chunk.LithiumHashPalette::create;
+    //private static final Configuration HASH_3BIT = new Configuration.Simple(HASH, 3);
+    private static final Configuration HASH_4BIT = new Configuration.Simple(HASH, 4);
+    private static final Configuration HASH_5BIT = new Configuration.Simple(HASH, 5);
+    private static final Configuration HASH_6BIT = new Configuration.Simple(HASH, 6);
+    private static final Configuration HASH_7BIT = new Configuration.Simple(HASH, 7);
+    private static final Configuration HASH_8BIT = new Configuration.Simple(HASH, 8);
+    // Leaf end - Lithium - faster hash palette
     public static final Palette.Factory SINGLE_VALUE_PALETTE_FACTORY = SingleValuePalette::create; // Paper - Anti-Xray
     private static final Palette.Factory LINEAR_PALETTE_FACTORY = LinearPalette::create;
     private static final Palette.Factory HASHMAP_PALETTE_FACTORY = HashMapPalette::create;
@@ -36,11 +45,14 @@ public abstract class Strategy<T> {
             public Configuration getConfigurationForBitCount(int bits) {
                 return (Configuration)(switch (bits) {
                     case 0 -> Strategy.ZERO_BITS;
-                    case 1, 2, 3, 4 -> Strategy.FOUR_BITS_LINEAR;
-                    case 5 -> Strategy.FIVE_BITS_HASHMAP;
-                    case 6 -> Strategy.SIX_BITS_HASHMAP;
-                    case 7 -> Strategy.SEVEN_BITS_HASHMAP;
-                    case 8 -> Strategy.EIGHT_BITS_HASHMAP;
+                    // Leaf start - Lithium - faster hash palette
+                    case 1, 2 -> Strategy.FOUR_BITS_LINEAR;
+                    case 3, 4 -> Strategy.HASH_4BIT;
+                    case 5 -> Strategy.HASH_5BIT;
+                    case 6 -> Strategy.HASH_6BIT;
+                    case 7 -> Strategy.HASH_7BIT;
+                    case 8 -> Strategy.HASH_8BIT;
+                    // Leaf end - Lithium - faster hash palette
                     default -> new Configuration.Global(this.globalPaletteBitsInMemory, bits);
                 });
             }
